driver/others/blas_server_win32.c

```diff
@@ -225,29 +229,38 @@ static DWORD WINAPI blas_thread_server(void *arg){
 #ifdef SMP_DEBUG
     fprintf(STDERR, "Server[%2ld] Waiting for Queue.\n", cpu);
 #endif
-
-    do {
-      action = WaitForMultipleObjects(2, handles, FALSE, INFINITE);
-    } while ((action != WAIT_OBJECT_0) && (action != WAIT_OBJECT_0 + 1));
-
-    if (action == WAIT_OBJECT_0 + 1) break;
+	// event raised when work is added to the queue
+	WaitForSingleObject(kickoff_event, INFINITE);
 
 #ifdef SMP_DEBUG
     fprintf(STDERR, "Server[%2ld] Got it.\n", cpu);
 #endif
 
-    EnterCriticalSection(&pool.lock);
+#if 1
```

Comment:
I wonder about this #if , which appears to disable newly introduced code in favor of something quite similar to the original ?

Reply:
Certainly, let me try to explain the approach that I took. Given the previous difficulties in passing the full set of runners, and that I can no longer run them on my fork, I wanted to take care to be somewhat conservative in the changes that I made, focusing on correctness while still improving performance. The code guarded by the #if 1 is very similar to what was there previously with the simplification of removing the pool structure below.

```
/* Thread server common information */
typedef struct{
  CRITICAL_SECTION lock;
  HANDLE filled;
  HANDLE killed;

  blas_queue_t	*queue;    /* Parameter Pointer */
  int		shutdown;  /* server shutdown flag */

} blas_pool_t;
 ```
The block of code with the lock is mostly house-keeping. The alternate path is a lock-free version that I am exploring, but it seemed too aggressive for a first PR. I will continue to test that an possibly suggest it in the future if we accept this PR and I find the lock-free version to be an additional improvement.

The important change in this PR is the removal of the CRITICAL_SECTION and EVENT from each blas_queue_t entry.These objects were created and destroyed in the processing of each queue element. That led to the unnecessasry creation/deletion of **millions** of Windows kernel objects over the course of a heavy workload like training a NN. That logic was replaced by use of an int volatile finished.

```
typedef struct blas_queue {

  void *routine;
  BLASLONG position;
  BLASLONG assigned;

  blas_arg_t *args;
  void *range_m;
  void *range_n;
  void *sa, *sb;

  struct blas_queue *next;

#if defined( __WIN32__) || defined(__CYGWIN32__) || defined(_WIN32) || defined(__CYGWIN__)
  // CRITICAL_SECTION lock;
  // HANDLE finish;
  volatile int finished;
#else
  pthread_mutex_t	 lock;
  pthread_cond_t	 finished;
#endif
```

Reply:
I see , thank you for the detailed explanation - I was primarily wondering if it would make sense to re-run CI with that conditional inverted to exercise the lockless implementation

Reply:
We could, but so far in my testing it has not conclusively shown itself to be faster. I wanted to explore that further before making the more significant change. I can remove the disabled CAS code path for now if that is problematic.

Thinking longer term I'd like to help **blas_server_win32.c** and **blas_server.c** converge towards a more common code base for easier maintainability. But that would be a more involved change and I'm reluctant to attempt it without more success and experience with the codebase.

Reply:
I don't think removing it is necessary, but perhaps adding a comment would be good (just in case it stays that way and someone does not have the luxury of tracking down this github exchange in the future for whatever reason). 
And yes, converging the "server" implementations would be nice but a bit challenging, given that it is legacy code with little to no documentation and all kinds of software depend on it,often without the end user even knowing.

Reply:
I just added the missing WIN_CASE definition, but did not add a comment. I also added a small bit of code to dynamically scale the number of threads so we don't have a ton of idle threads that can spuriously wake up and consume resources. I will wait to see how the build clears before I add a comment.

Reply:
It looks like the 2 builds that failed were OSX and Linux builds so should not be affected by the code in my PR. Does that seem correct?

Reply:
Yes, seems they just fell over trying to install dependencies for the build... happens occasionally